core/src/nodes/collection.rs

```diff
@@ -143,65 +143,85 @@ where TOutEvent: fmt::Debug,
 
 /// Event that happens when we reach a node.
 #[must_use = "The node reached event is used to accept the newly-opened connection"]
-pub struct CollectionReachEvent<'a, TInEvent, TOutEvent, THandler, TReachErr, THandlerErr, TUserData, TPeerId = PeerId> {
-    /// Peer id we connected to.
-    peer_id: TPeerId,
+pub struct CollectionReachEvent<'a, TInEvent, TOutEvent, THandler, TReachErr, THandlerErr, TUserData, TConnInfo = PeerId, TPeerId = PeerId> {
+    /// Information about the connection, or `None` if it's been extracted.
+    conn_info: Option<TConnInfo>,
     /// The task id that reached the node.
     id: TaskId,
     /// The `CollectionStream` we are referencing.
-    parent: &'a mut CollectionStream<TInEvent, TOutEvent, THandler, TReachErr, THandlerErr, TUserData, TPeerId>,
+    parent: &'a mut CollectionStream<TInEvent, TOutEvent, THandler, TReachErr, THandlerErr, TUserData, TConnInfo, TPeerId>,
 }
 
-impl<'a, TInEvent, TOutEvent, THandler, TReachErr, THandlerErr, TUserData, TPeerId>
-    CollectionReachEvent<'a, TInEvent, TOutEvent, THandler, TReachErr, THandlerErr, TUserData, TPeerId>
-where
-    TPeerId: Eq + Hash + Clone,
+impl<'a, TInEvent, TOutEvent, THandler, TReachErr, THandlerErr, TUserData, TConnInfo, TPeerId>
+    CollectionReachEvent<'a, TInEvent, TOutEvent, THandler, TReachErr, THandlerErr, TUserData, TConnInfo, TPeerId>
 {
-    /// Returns the peer id of the node that has been reached.
-    #[inline]
-    pub fn peer_id(&self) -> &TPeerId {
-        &self.peer_id
+    /// Returns the information of the connection.
+    pub fn connection_info(&self) -> &TConnInfo {
+        self.conn_info.as_ref().expect("conn_info is always Some when the object is alive; QED")
+    }
+
+    /// Returns the identity of the node we connected to.
+    pub fn peer_id(&self) -> &TPeerId
```

Comment:
How about removing `peer_id()`?

Reply:
I had it for convenience, as a "why not have it" item, but I'll remove it.

Reply:
It is there for convenience and is used a lot in `raw_swarm`. Can I keep it? :pray:

Reply:
Of course you can keep it :smile:. It was just a thought.